### ISSUE TITLE
Gait quality slowly collapses on humanoid locomotion (reward drifts down smoothly)

### ISSUE THREAD

**Issue by ningyuanz** [younggyoseo/FastTD3#26]
Hi fast_td3 authors—thank you for the great work and for sharing this implementation! It’s been extremely helpful for my locomotion experiments, and I appreciate the time you put into maintaining it.

**Summary**
With fast_td3, training starts well but then reward (and gait quality) slowly degrades over time. The collapse is very smooth (no spikes). TD3 frequently converges to unnatural gaits (tiptoe, foot pronation). At the best point before degradation, gait is acceptable but still worse than PPO (larger upper-body swing, lower leg lift).

I also observed that the Q-loss drops from ~10 → ~2, then rises to ~4 and plateaus. The reward peak occurs after the Q-loss bottoms out and has already plateaued.

I tried typical TD3 stabilizers, but the issue persists. Things I have done:
1) Increase policy_noise up to 0.25;
2) Increase policy_frequency up to 4;
3) Decrese tau to 0.005;
4) Decrease actor_learning_rate to 1e-4 while keep critic_learning_rate at 3e-4;
5) Tune rewards and v_min/v_max accordingly;
6) Try OU noise instead of Gaussian;
7) Tune batch_size and model size according to this blog: https://value-scaling.github.io/

**Environment**: IsaacGym environment with a full-sized humanoid.
**Task**: full-size humanoid velocity command following (track v_x, v_y, yaw_rate)
**Baseline**: PPO trains reliably with natural gait and good tracking using the same set of rewards (weights are tuned specifically for fast td3 as suggested in the paper)

I am kinda getting stuck here. Could this smooth reward drift toward toe-walking/pronation be a known failure mode with fast_td3 on contact-rich humanoid tasks, and if so, could you share any recommended mitigations? Or would you be able to suggest preferred settings?

Thanks!

**Comment by ningyuanz** [comment]
Hi @younggyoseo, after a lot of trial and error, I finally adapted FastTD3 to produce policies with gait and task rewards comparable to PPO. What I did:
1) Lower regularization rewards weights (action rate, torques, etc) to be approximately 20% of PPO's values;
2) Switched to OU noise or gSDE instead of Gaussian — **this was probably the most important one.**
3) Higher num_steps;
4) More reasonable action buffering when action_bounds != 1;

**Comment by younggyoseo** [comment]
Amazing! I guess that is aligned with the [Antonin's observation where exploration was quite effective for making SAC better?](https://araffin.github.io/post/tune-sac-isaac-sim/)

**Comment by ningyuanz** [comment]
Yes—both the gSDE idea and using larger num_steps came from that blog post. In my runs, though, FastTD3 ends up on par with PPO in final performance.

One change that turned out to be important for both speed and final performance:

Store the executed (clamped) actions in the replay buffer when action_bounds != 1.0. This keeps the data the critic/actor learn from consistent with what the environment actually saw. For example in isaaclab_env.py we have:

```
if self.action_bounds is not None:
    actions = torch.clamp(actions, -1.0, 1.0) * self.action_bounds    # <--- Store the actions after clamp 
```

Also, make sure the action scale is consistent in q net and actor inputs.
